fish-rust/src/io.rs

```diff
@@ -894,8 +947,8 @@ impl OutputStream for BufferedOutputStream {
 
 pub struct IoStreams<'a> {
     // Streams for out and err.
-    pub out: &'a dyn OutputStream,
-    pub err: &'a dyn OutputStream,
+    pub out: &'a mut dyn OutputStream,
```

Comment:
https://github.com/fish-shell/fish-shell/commit/d6b41b33f4c888bbce3cba1e14e3dbf9dadc0afd#r124997448

> I wrote: OutputStream should really not use dynamic dispatch IMO, see my previous comment https://github.com/fish-shell/fish-shell/pull/9916#issuecomment-1655563903
> 
> > Aside: regarding IoStreams, I think it is better to avoid object-safety and use an OutputStream-enum that delegates, like [StringMatcher](https://github.com/fish-shell/fish-shell/blob/2ec36338f2b1f21f4074ff7184c37619a4a61e69/fish-rust/src/builtins/string/match.rs#L157) from string match. Then we can avoid the API breakage of no generics. Using dyn would be fine if we used [io::Write](https://doc.rust-lang.org/stable/std/io/trait.Write.html) + [write!(streams.out, output)](https://doc.rust-lang.org/stable/std/macro.write.html), but that's a much larger change that I don't think we want right now. I think that a delegating enum is the closest we have to subclassing from C++, even if it is a bit of boilerplate.
> 
> I think you can use generics if you implement something directly on `impl dyn OutputStream`, but I'm not very familiar with that, you generally want to avoid dynamic dispatch if possible due to requiring object-safety

> @krobelus responded: 
> > Then we can avoid the API breakage of no generics.
> yeah good point

Reply:
We could use both `io::Write` to skip private-encode-undo-ing[^2] and `fmt::Write` to take possibly-private-encoded UTF8[^1], and undo the private-encoding before writing it to the stream. (but that is for a potential UTF8-swap)

[^2]: That sounds mildly risky
[^1]: `WString` currently has a dual purpose of UTF32 + private-encoded